File: week1_baseline/ruby/11_tui/lib/boukensha/mcp_servers/file_system_server.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# MCP server exposing the standard file-system tools, sandboxed to a single
# root directory. Run as a stdio subprocess by
# lib/boukensha/tools/file_system.rb — the root directory is passed via the
# BOUKENSHA_MCP_FS_ROOT environment variable, since that's how the parent
# process (mcp_client.rb) launches it.
#
# Tool logic here is identical to the pre-MCP tools/file_system.rb
# implementation — only the transport changed (local closures -> an MCP
# server the framework talks to over stdio).

require "bundler/setup" if ENV["BUNDLE_GEMFILE"]
require "mcp"
require "fileutils"

ROOT = File.expand_path(ENV.fetch("BOUKENSHA_MCP_FS_ROOT"))

def resolve(path)
  absolute = File.expand_path(path.to_s, ROOT)
  if absolute == ROOT || absolute.start_with?("#{ROOT}/")
    absolute
  else
    "error: path '#{path}' escapes the working directory"
  end
end

def oops(msg) = "error: #{msg}"

def text_response(str)
  MCP::Tool::Response.new([{ type: "text", text: str }])
end

server = MCP::Server.new(name: "boukensha-file-system")

server.define_tool(
  name: "pwd",
  description: "Return the working directory — the root that all file paths are relative to."
) do |server_context:, **_args|
  text_response(ROOT)
end

server.define_tool(
  name: "list_directory",
  description: "List files and subdirectories at a path relative to the working directory. Defaults to the working directory itself.",
  input_schema: { properties: { path: { type: "string", description: "Relative path to list (default '.')" } } }
) do |server_context:, **args|
  path   = args[:path] || "."
  target = resolve(path)
  next text_response(target) if target.start_with?("error:")
  next text_response(oops("'#{path}' is not a directory")) unless File.directory?(target)

  entries = Dir.entries(target)
               .reject { |e| e == "." || e == ".." }
               .sort
               .map { |name| File.directory?(File.join(target, name)) ? "#{name}/" : name }

  text_response(entries.empty? ? "(empty)" : entries.join("\n"))
end

server.define_tool(
  name: "read_file",
  description: "Read and return the full contents of a file. Path is relative to the working directory.",
  input_schema: { properties: { path: { type: "string", description: "Relative path to the file" } }, required: ["path"] }
) do |server_context:, **args|
  target = resolve(args[:path])
  next text_response(target) if target.start_with?("error:")
  next text_response(oops("'#{args[:path]}' is not a file")) unless File.file?(target)

  begin
    text_response(File.read(target))
  rescue => e
    text_response(oops(e.message))
  end
end

server.define_tool(
  name: "write_file",
  description: "Write content to a file, creating it (and any missing parent directories) if needed, overwriting if it exists. Path is relative to the working directory.",
  input_schema: {
    properties: {
      path:    { type: "string", description: "Relative path to the file" },
      content: { type: "string", description: "Text content to write" }
    },
    required: ["path", "content"]
  }
) do |server_context:, **args|
  target = resolve(args[:path])
  next text_response(target) if target.start_with?("error:")

  begin
    FileUtils.mkdir_p(File.dirname(target))
    File.write(target, args[:content])
    rel = target.delete_prefix("#{ROOT}/")
    text_response("ok: wrote #{args[:content].bytesize} bytes to #{rel}")
  rescue => e
    text_response(oops(e.message))
  end
end

server.define_tool(
  name: "delete_file",
  description: "Delete a file. Directories are not deleted. Path is relative to the working directory.",
  input_schema: { properties: { path: { type: "string", description: "Relative path to the file to delete" } }, required: ["path"] }
) do |server_context:, **args|
  target = resolve(args[:path])
  next text_response(target) if target.start_with?("error:")
  next text_response(oops("'#{args[:path]}' is not a file")) unless File.file?(target)

  begin
    File.delete(target)
    text_response("ok: deleted #{args[:path]}")
  rescue => e
    text_response(oops(e.message))
  end
end

server.define_tool(
  name: "search_files",
  description: "Search for a text pattern (literal string or Ruby regex) across all files in the working directory tree. Returns matching lines in 'path:line_number:content' format.",
  input_schema: {
    properties: {
      pattern: { type: "string", description: "The text or regex pattern to search for" },
      path:    { type: "string", description: "Subdirectory or file to search within (default '.' = entire working directory)" },
      glob:    { type: "string", description: "File glob to restrict which files are searched, e.g. '*.rb' (default '*')" }
    },
    required: ["pattern"]
  }
) do |server_context:, **args|
  path, glob = args[:path] || ".", args[:glob] || "*"
  target = resolve(path)
  next text_response(target) if target.start_with?("error:")

  file_glob = File.file?(target) ? target : File.join(target, "**", glob)

  begin
    regex = Regexp.new(args[:pattern])
  rescue RegexpError => e
    next text_response(oops("invalid pattern: #{e.message}"))
  end

  matches = []
  Dir.glob(file_glob).sort.each do |file|
    next unless File.file?(file)
    rel = file.delete_prefix("#{ROOT}/")
    begin
      File.foreach(file).with_index(1) do |line, lineno|
        matches << "#{rel}:#{lineno}:#{line.chomp}" if line.match?(regex)
      end
    rescue => e
      matches << "#{rel}: error reading file: #{e.message}"
    end
  end

  text_response(matches.empty? ? "no matches" : matches.join("\n"))
end

MCP::Server::Transports::StdioTransport.new(server).open

File: week1_baseline/python/11_tui/boukensha/mcp_servers/file_system_server.py
"""MCP server exposing the standard file-system tools, sandboxed to a
single root directory. Run as a stdio subprocess by
boukensha/tools/file_system.py — the root directory is passed via the
BOUKENSHA_MCP_FS_ROOT environment variable rather than a CLI argument or
config file, since that's how the parent process (boukensha/mcp_client.py's
McpClient) launches it.

Tool logic here is identical to the pre-MCP boukensha/tools/file_system.py
implementation — only the transport changed (local Python closures ->
an MCP server the framework talks to over stdio).
"""

import glob as _glob
import os
import re
from typing import Annotated

from pydantic import Field

from mcp.server.fastmcp import FastMCP

ROOT = os.path.abspath(os.environ["BOUKENSHA_MCP_FS_ROOT"])

mcp = FastMCP("boukensha-file-system")


def _resolve(path):
    path = str(path)
    joined = path if os.path.isabs(path) else os.path.join(ROOT, path)
    absolute = os.path.normpath(joined)
    if absolute == ROOT or absolute.startswith(ROOT + os.sep):
        return absolute
    return f"error: path '{path}' escapes the working directory"


def _oops(msg):
    return f"error: {msg}"


@mcp.tool()
def pwd() -> str:
    """Return the working directory — the root that all file paths are relative to."""
    return ROOT


@mcp.tool()
def list_directory(
    path: Annotated[str, Field(description="Relative path to list (default '.')")] = ".",
) -> str:
    """List files and subdirectories at a path relative to the working directory. Defaults to the working directory itself."""
    target = _resolve(path)
    if target.startswith("error:"):
        return target
    if not os.path.isdir(target):
        return _oops(f"'{path}' is not a directory")

    entries = sorted(os.listdir(target))
    entries = [e + "/" if os.path.isdir(os.path.join(target, e)) else e for e in entries]
    return "\n".join(entries) if entries else "(empty)"


@mcp.tool()
def read_file(
    path: Annotated[str, Field(description="Relative path to the file")],
) -> str:
    """Read and return the full contents of a file. Path is relative to the working directory."""
    target = _resolve(path)
    if target.startswith("error:"):
        return target
    if not os.path.isfile(target):
        return _oops(f"'{path}' is not a file")
    try:
        with open(target, "r") as f:
            return f.read()
    except OSError as e:
        return _oops(str(e))


@mcp.tool()
def write_file(
    path: Annotated[str, Field(description="Relative path to the file")],
    content: Annotated[str, Field(description="Text content to write")],
) -> str:
    """Write content to a file, creating it (and any missing parent directories) if needed, overwriting if it exists. Path is relative to the working directory."""
    target = _resolve(path)
    if target.startswith("error:"):
        return target
    try:
        os.makedirs(os.path.dirname(target), exist_ok=True)
        with open(target, "w") as f:
            f.write(content)
        rel = os.path.relpath(target, ROOT)
        return f"ok: wrote {len(content.encode('utf-8'))} bytes to {rel}"
    except OSError as e:
        return _oops(str(e))


@mcp.tool()
def delete_file(
    path: Annotated[str, Field(description="Relative path to the file to delete")],
) -> str:
    """Delete a file. Directories are not deleted. Path is relative to the working directory."""
    target = _resolve(path)
    if target.startswith("error:"):
        return target
    if not os.path.isfile(target):
        return _oops(f"'{path}' is not a file")
    try:
        os.remove(target)
        return f"ok: deleted {path}"
    except OSError as e:
        return _oops(str(e))


@mcp.tool()
def search_files(
    pattern: Annotated[str, Field(description="The text or regex pattern to search for")],
    path: Annotated[str, Field(description="Subdirectory or file to search within (default '.' = entire working directory)")] = ".",
    glob: Annotated[str, Field(description="File glob to restrict which files are searched, e.g. '*.py' (default '*')")] = "*",
) -> str:
    """Search for a text pattern (literal string or Python regex) across all files in the working directory tree. Returns matching lines in 'path:line_number:content' format."""
    target = _resolve(path)
    if target.startswith("error:"):
        return target

    search_is_file = os.path.isfile(target)
    file_glob = target if search_is_file else os.path.join(target, "**", glob)

    try:
        regex = re.compile(pattern)
    except re.error as e:
        return _oops(f"invalid pattern: {e}")

    matches = []
    for file in sorted(_glob.glob(file_glob, recursive=True)):
        if not os.path.isfile(file):
            continue
        rel = os.path.relpath(file, ROOT)
        try:
            with open(file, "r", errors="replace") as f:
                for lineno, line in enumerate(f, start=1):
                    if regex.search(line):
                        matches.append(f"{rel}:{lineno}:{line.rstrip(chr(10))}")
        except OSError as e:
            matches.append(f"{rel}: error reading file: {e}")

    return "\n".join(matches) if matches else "no matches"


if __name__ == "__main__":
    mcp.run(transport="stdio")
